src/Pact/Native/Trans/Types.hs

```diff
@@ -0,0 +1,263 @@
+  c'mpz_get_str :: Ptr CChar -> CInt -> Mpz_t -> IO (Ptr CChar)
+-}
+
+foreign import ccall "__gmpq_init"
+  c'mpq_init :: Mpq_t -> IO ()
+
+foreign import ccall "__gmpq_clear"
+  c'mpq_clear :: Mpq_t -> IO ()
+
+foreign import ccall "__gmpq_set_str"
+  c'mpq_set_str :: Mpq_t -> Ptr CChar -> CInt -> IO ()
+
+foreign import ccall "__gmpq_get_str"
+  c'mpq_get_str :: Ptr CChar -> CInt -> Mpq_t -> IO (Ptr CChar)
+
+foreign import ccall "mpfr_init"
+  c'mpfr_init :: Mpfr_t -> IO ()
+
+foreign import ccall "mpfr_set_default_prec"
+  c'mpfr_set_default_prec :: CInt -> IO ()
+
+foreign import ccall "mpfr_clear"
+  c'mpfr_clear :: Mpfr_t -> IO ()
+
+foreign import ccall "mpfr_set_str"
+  c'mpfr_set_str :: Mpfr_t -> Ptr CChar -> CInt -> CInt -> IO ()
+
+foreign import ccall "mpfr_set_q"
+  c'mpfr_set_q :: Mpfr_t -> Mpq_t -> CInt -> IO ()
+
+foreign import ccall "mpfr_get_q"
+  c'mpfr_get_q :: Mpq_t -> Mpfr_t -> IO ()
+
+foreign import ccall "mpfr_div"
+  c'mpfr_div :: Mpfr_t -> Mpfr_t -> Mpfr_t -> CInt -> IO ()
+
+foreign import ccall "mpfr_pow"
+  c'mpfr_pow :: Mpfr_t -> Mpfr_t -> Mpfr_t -> CInt -> IO ()
+
+foreign import ccall "mpfr_log"
+  c'mpfr_log :: Mpfr_t -> Mpfr_t -> CInt -> IO ()
+
+foreign import ccall "mpfr_log2"
+  c'mpfr_log2 :: Mpfr_t -> Mpfr_t -> CInt -> IO ()
+
+foreign import ccall "mpfr_log10"
+  c'mpfr_log10 :: Mpfr_t -> Mpfr_t -> CInt -> IO ()
+
+foreign import ccall "mpfr_exp"
+  c'mpfr_exp :: Mpfr_t -> Mpfr_t -> CInt -> IO ()
+
+foreign import ccall "mpfr_exp2"
+  c'mpfr_exp2 :: Mpfr_t -> Mpfr_t -> CInt -> IO ()
+
+foreign import ccall "mpfr_exp10"
+  c'mpfr_exp10 :: Mpfr_t -> Mpfr_t -> CInt -> IO ()
+
+foreign import ccall "mpfr_sqrt"
+  c'mpfr_sqrt :: Mpfr_t -> Mpfr_t -> CInt -> IO ()
+
+foreign import ccall "mpfr_snprintf"
+  c'mpfr_snprintf :: Ptr CChar -> CInt -> Ptr CChar -> CInt -> Mpfr_t -> IO ()
+
+withTempz :: (Mpz_t -> IO a) -> IO a
+withTempz k = alloca $ \x ->
+  bracket_ (c'mpz_init x) (c'mpz_clear x) (k x)
+
+withTempq :: (Mpq_t -> IO a) -> IO a
+withTempq k = alloca $ \x ->
+  bracket_ (c'mpq_init x) (c'mpq_clear x) (k x)
+
+withTemp :: (Mpfr_t -> IO a) -> IO a
+withTemp k = alloca $ \x ->
+  bracket_ (c'mpfr_init x) (c'mpfr_clear x) (k x)
+
+dec2Mpfr :: Decimal -> (Mpfr_t -> IO a) -> IO a
+dec2Mpfr d k =
+  withCString (show (numerator r) ++ "/" ++ show (denominator r)) $ \r' ->
+  withTempq $ \q ->
+  withTemp $ \x -> do
+    c'mpq_set_str q r' 10
+    c'mpfr_set_q x q rounding
+    k x
+  where
+  r = toRational d
+
+mpfr2Dec :: Mpfr_t -> IO (TransResult Decimal)
+mpfr2Dec m =
+  withTempq $ \q -> do
+    c'mpfr_get_q q m
+    out <- c'mpq_get_str nullPtr 10 q
+    buf <- peekCString out
+    free out
+    let val = case break (== '/') buf of
+            (before, []) -> read before % 1
+            (before, _:after) -> read before % read after
+    pure $ TransNumber $ fromRational val
+
+mpfr_arity1
+  :: (Mpfr_t -> Mpfr_t -> CInt -> IO ()) -> Decimal -> TransResult Decimal
+mpfr_arity1 f x = unsafePerformIO $
```

Comment:
It seems like the use of unsafePerformDupableIO would be safe here, which would improve performance.

```suggestion
mpfr_arity1 f x = unsafeDupablePerformIO $
```

Reply:
The performance gain would be so exceedingly minimal, compared to the cost of the operations themselves, that I'd rather be naïvely conservative.